lightning/src/routing/router.rs

```diff
@@ -216,84 +379,213 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+
+			// It is tricky to substract $next_hops_fee_msat from available liquidity here.
+			// It may be misleading because we might later choose to reduce the value transferred
+			// over these channels, and the channel which was insufficient might become sufficient.
+			// Worst case: we drop a good channel here because it can't cover the high following
+			// fees caused by one expensive channel, but then this channel could have been used
+			// if the amount being transferred over this path is lower.
+			// We do this for now, but this is a subject for removal.
+			if let Some(available_value_contribution_msat) = available_liquidity_msat.checked_sub($next_hops_fee_msat) {
+
+				// Routing Fragmentation Mitigation heuristic:
+				//
+				// Routing fragmentation across many payment paths increases the overall routing
+				// fees as you have irreducible routing fees per-link used (`fee_base_msat`).
+				// Taking too many smaller paths also increases the chance of payment failure.
+				// Thus to avoid this effect, we require from our collected links to provide
+				// at least a minimal contribution to the recommended value yet-to-be-fulfilled.
+				//
+				// This requirement is currently 5% of the remaining-to-be-collected value.
+				// This means as we successfully advance in our collection,
+				// the absolute liquidity contribution is lowered,
+				// thus increasing the number of potential channels to be selected.
+
+				// Derive the minimal liquidity contribution with a ratio of 20 (5%, rounded up).
+				let minimal_value_contribution_msat: u64 = (recommended_value_msat - already_collected_value_msat + 19) / 20;
+				// Verify the liquidity offered by this channel complies to the minimal contribution.
+				let contributes_sufficient_value = available_value_contribution_msat >= minimal_value_contribution_msat;
+
+				let value_contribution_msat = cmp::min(available_value_contribution_msat, $next_hops_value_contribution);
+				// Includes paying fees for the use of the following channels.
+				let amount_to_transfer_over_msat: u64 = match value_contribution_msat.checked_add($next_hops_fee_msat) {
+					Some(result) => result,
+					// Can't overflow due to how the values were computed right above.
+					None => unreachable!(),
+				};
+
+				// If HTLC minimum is larger than the whole value we're going to transfer, we shouldn't bother
+				// even considering this channel, because it won't be useful.
+				// TODO: Explore simply adding fee to hit htlc_minimum_msat
+				// This is separate from checking amount_to_transfer_over_msat, which also should be
+				// lower-bounded by htlc_minimum_msat. Since we're choosing it as maximum possible,
+				// this channel should just be skipped if it's not sufficient.
+				// amount_to_transfer_over_msat can be both larger and smaller than final_value_msat,
+				// so this can't be derived.
+				let final_value_satisfies_htlc_minimum_msat = final_value_msat >= $directional_info.htlc_minimum_msat;
```

Comment:
I don't understand why failing this check should discard the channel to be selected.

On link N, htlc_minimum_msat is superior to final_value_msat but it doesn't matter as `amount_to_transfer_over_msat` is superior by including hops {N+1, ... N+i} fees. It will be reduced to either final_value_msat or a good-enough contribution when it reaches destination.

Reply:
I agree, this check probably can be dropped. We could have checked against final_value_msat+$next_hops_fee_msat,
but that becomes too similar to amount_to_transfer_over_msat check, which is already there.